source/agora/test/RestoreSCPState.d

```diff
@@ -0,0 +1,178 @@
+/*******************************************************************************
+
+    Tests restoring SCP Envelope state on restart
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.RestoreSCPState;
+
+version (unittest):
+
+import agora.common.Config;
+import agora.common.Serializer;
+import agora.common.Task;
+import agora.common.Hash;
+import agora.common.crypto.Key;
+import agora.consensus.protocol.Nominator;
+import agora.consensus.data.Block;
+import agora.consensus.data.ConsensusParams;
+import agora.consensus.data.Transaction;
+import agora.consensus.SCPEnvelopeStore;
+import agora.node.Ledger;
+import agora.node.Validator;
+import agora.network.Clock;
+import agora.network.NetworkManager;
+import agora.test.Base;
+
+import geod24.Registry;
+
+import scpd.Cpp;
+import scpd.types.Stellar_SCP;
+
+import core.stdc.time;
+import core.thread;
+
+/// Class containing gshared store for SCPEnvelopeStoreTest
+public class TestSCPEnvelopeStore : SCPEnvelopeStore
+{
+    __gshared const(SCPEnvelope)[] store;
+
+    this ()
+    {
+        super(":memory:");
+    }
+
+    public override bool add (const ref SCPEnvelope envelope) @trusted nothrow
+    {
+        store ~= envelope;
+        return true;
+    }
+
+    public override void removeAll () @trusted nothrow
+    {
+        // no-op
+    }
+
+    public override int opApply (scope int delegate(const ref SCPEnvelope) dg) @trusted
+    {
+        foreach (ref env; store)
+        {
+            if (auto ret = dg(env))
+                return ret;
+        }
+        return 0;
+    }
+
+    public override size_t length () @trusted
+    {
+        return store.length;
+    }
+}
+
+/// A test to store SCP state and recover SCP state
+unittest
+{
+    __gshared bool Checked = false;
+
+    static class ReNominator : TestNominator
+    {
+        /// Ctor
+        public this (immutable(ConsensusParams) params, Clock clock,
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman, string data_dir, ulong txs_to_nominate)
+        {
+            super(params, clock, network, key_pair, ledger, taskman, data_dir,
+                txs_to_nominate);
+        }
+
+        ///
+        override void restoreSCPState ()
+        {
+            // on first boot the envelope store will be empty,
+            // but on restart it should contain two envelopes
+            // (nominate & ballot protocol messages)
+            if (Checked)
+            {
+                assert(TestSCPEnvelopeStore.store.length == 2);
+                Checked = true;
```

Comment:
There was some misunderstanding. I wanted you to use two booleans. However we can use a single boolean too. But, we need to change the code once more.

```suggestion
                Checked = false;
```

Reply:
Yes, that's right.